### ISSUE TITLE
Support `Decimal128` / `Decimal256` in json inference

### ISSUE THREAD

**Issue by kskalski** [apache/arrow-rs#4856]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I'm processing jsons with some exceptionally large numbers provided as json `number`. They exceed both `int64` and exactly represented `float64`.

Currently json inference fallback to `float64`, which causes reader to lose exactness when parsing values.

**Describe the solution you'd like**
I think since Rust `arrow` now supports decimal data type, it could be used also for json inference (parsing already works if I enforce a proper data type).

I'm not sure which precision/scale should be inferred in this case, possibly inference should track necessary scale and assume `(max precision - necessary scale, necessary scale)`.

**Describe alternatives you've considered**
One way to work around that would be to use strings, since `json::ReaderBuilder` with `.with_coerce_primitive(true)` allows to parse interchangebly strings and numbers.
This might actually still be a valid solution if we need to support numbers that are beyond even `decimal256`.

**Additional context**
```rust
   const DATA_BIG_NUM: &str = r#"{"x": 1, "y": "01"}
    {"x": 2000000000, "y": "02"}
    {"x": 102174585645880460651957762, "y": "03"}
    {"x": 102174585645880460651957762102, "y": "04"}"#;

    #[test]
    fn test_json_infers_coerce() {
        let mut cursor = std::io::Cursor::new(&DATA_BIG_NUM);
        let schema = infer_json_schema_from_seekable(&mut cursor, Some(3)).unwrap();
        let mut json = json::ReaderBuilder::new(schema.into()).build(cursor).unwrap();

        let batch = json.next().unwrap().unwrap();
        // Inference falls into float64 datatype, which is lossy as demonstrated below
        assert_eq!(DataType::Decimal128(30, 0), *batch.schema().fields()[0].data_type());
        assert_eq!(
            102174585645880460651957762i128,
            Decimal128Array::from(batch.column(0).to_data()).value(2)
        );
    }
```

**Comment by tustvold** [comment]
The challenge here is determining what the correct precision/scale is for the input. 

Whilst you could parse all the values, this is potentially:

* Expensive
* Incomplete if only viewing a subset of the data

Perhaps you might be able to find a system that does something similar to this already that we could use as a reference point on what the "expected" behaviour actually is? Whilst the situation is less terrible than CSV, which lacks any specification, there isn't universal agreement on what JSON's numeric type actually is, at least historically it was strictly f64...

**Comment by kskalski** [comment]
The [JSON spec](https://www.rfc-editor.org/rfc/rfc8259.html#section-6)(JSON is pretty vague on the semantics of "numbers", focusing on syntax and only suggesting an "interoperable" range of numbers, which follows `float64` implementation (`[-(2^53)+1, (2^53)-1]`). So technically "number" syntax is valid also for numbers without exact `f64` representation, it's just implied that many systems will fail to read them precisely.

In the wild I think the common approach is:
- for parsers to accept any numbers with conforming syntax, interpreting them with maximum available/configured/implemented precision, e.g. `serde_json` will automatically move from using `u64` to `f64`, but also allows compiling with `arbitrary_precision` feature, in which case values are stored in memory as strings and conversion to numeric types are done on the fly
- for serializers to fallback into json strings when number to be serialized falls outside of "interoperable" range
- previous point imposes additional nice-to-have feature on parsers to permit "numeric strings" when they expect numbers

Clearly picking wider storage type than required is a performance penalty and possibly there should be a choice whether user wants to pay it. I would argue that whenever we detect data point that *requires* wider type for exact representation, the default behavior should be to preserve it or signal error, so for inference it means switching to wider type, e.g. `int32` -> `int64` -> `decimal128` -> `decimal256` -> possibly down to `string`. The latter is a proper choice when user doesn't want to lose precision + enables `coerce_primitive` flag in the reader.
Additional questions arise here:
* whether inference should perform a 100% precise detection of representable type boundaries, since that really requires more complex logic than parsing numbers to a known target representation.
* if the inference logic should differ when max number of rows is specified or not, there is temptation that limited inference should be more lenient on picking a type, while unlimited inference could assume that it has seen all the data (so e.g. it could use `int32` instead of `int64` when possible)

The current behaviour of lossy parsing to `f64`, which to a large degree is possible for almost any number, could be kept under a flag - essentially it implements a specific use-case where user prefers performance more than correctness.

**Comment by tustvold** [comment]
I think my comment was open to misinterpretation, my question does not concern whether we should widen the type based on observed data, I think we definitely should. The challenge is the decimal types are a lot more than just wider integers, they also have notions of precision and scale, as they're designed to encode fixed precision decimals. Whilst we _could_ just infer a scale of 0 and the maximum precision, this doesn't immediately seem right.... 

This brings me back to my original question, what do other implementations do here, e.g. what does arrow-cpp/pyarrow do?

**Comment by kskalski** [comment]
I see, right. Picking too small scale could cause trouble easily, maybe this datatype makes sense in inference only for integers (no floating point).
Right now [pyarrow](https://arrow.apache.org/docs/dev/python/json.html) and [C++](https://arrow.apache.org/docs/dev/cpp/json.html) seem to have very similar mode to `arrow-rs` where inference only goes between `int64` and `float64`

Maybe an option to infer string for unrepresentable numbers would make more sense first then.

**Comment by tustvold** [comment]
> Maybe an option to infer string for unrepresentable numbers would make more sense first then.

Makes sense to me, and would let the user decide how they want to handle this
